### DISCUSSION TITLE
Table column ordering

### DISCUSSION THREAD

**Discussion by gar1t** [nushell/nushell#10301] [Q&A]
When rendering data like this:

    > [{"z": 123, "a": 345}, {"z": 456, "a": 678}]

the table seems to consistently order the columns as `| z | a |`

Is this a reliable interface? I.e. the table uses the flag order specified in the first record to order its columns?

In looking over https://www.nushell.sh/book/working_with_tables.html I didn't see anything specified beyond the interface for adding columns (which presumably add in left-to-right order).

**Comment by amtoine** [comment]
yes, i think the columns will be shown with the same order as in the first record of the table :+1: 

this also means that the *order* of the columns in the data does not matter, e.g.
```nushell
[{"z": 123, "a": 345}, {"a": 678, "z": 456}]
```
will give the same table as your example above :ok_hand:

**Reply by amtoine** [reply]
the two tables are indeed truely the same :+1: 
```nushell
use std assert

assert equal [{"z": 123, "a": 345}, {"z": 456, "a": 678}] [{"z": 123, "a": 345}, {"a": 678, "z": 456}]
```

**Comment by gar1t** [comment]
It'd be very surprising if Nushell treated records as different values based on the order the flags appear in an encoded form.

The behavior of the table *presentation* over stdout though I think should be explicitly defined, if it isn't already. There's no hint in the docs (at least the link I posted above) on what that behavior is intended to be. E.g. a `--columns` flag to the `table` command could be used to explicitly order the columns presented - and even to filter columns, which I think would be useful.

As a somewhat related topic, I see that `ls` columns vary based on flags (e.g. `-l`). This determines not only the columns that show up in a table presentation, but also the presence/absence of table values.

This fails:

    > ls | where group == garrett

To filter by group you need to specify `-l`.

This scheme of coupling the data structure with the presentation I think limits the applications of Nushell's excellent command interoperability.

I may be missing something fundamental here though!

**Comment by amtoine** [comment]
the default behaviour of table column ordering, i.e. follow the order of the first record, might be missing from the documentation? :+1: 

if you want to order the columns as you wish, you can use the `select` command
```nushell
ls | select type modified size name
```

currently, we do not have support for complex input / output types, i.e. you can only specify `table` but not `table<foo: ..., bar: ...>` :thinking: 
this is what the signature of `ls` look like
```
  #┬─input─┬output
  0│nothing│table
  ─┴───────┴──────
```
but we could imagine having something more informative, such as
```
  #┬─input─┬option┬────────────────────────────────────────────────────────────────────────────────────────output──────────────────────────────────────────────────────────────────────────────────────────────────────────
  0│nothing│      │table<name: string, type: string, size: filesize, modified: date>
  1│nothing│--long│table<name: string, type: string, target: nothing, readonly: bool, mode: string, num_links: int, inode: int, user: string, group: string, size: filesize, created: date, accessed: date, modified: date>
  ─┴───────┴──────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```
or maybe easier to read
```
#┬─input─┬option┬─────────output─────────
0│nothing│      │─────┬─────────────────
 │       │      │     │────────┬────────
 │       │      │table│name    │string
 │       │      │     │type    │string
 │       │      │     │size    │filesize
 │       │      │     │modified│date
 │       │      │     │────────┴────────
 │       │      │─────┴─────────────────
1│nothing│--long│─────┬──────────────────
 │       │      │     │─────────┬────────
 │       │      │table│name     │string
 │       │      │     │type     │string
 │       │      │     │target   │nothing
 │       │      │     │readonly │bool
 │       │      │     │mode     │string
 │       │      │     │num_links│int
 │       │      │     │inode    │int
 │       │      │     │user     │string
 │       │      │     │group    │string
 │       │      │     │size     │filesize
 │       │      │     │created  │date
 │       │      │     │accessed │date
 │       │      │     │modified │date
 │       │      │     │─────────┴────────
 │       │      │─────┴──────────────────
─┴───────┴──────┴────────────────────────
```

**Reply by amtoine** [reply]
as you can see, this will need some design :yum:

**Comment by gar1t** [comment]
I'm only speculating but I would imagine this to a property of the *data* that's passed along - some UI/presentation decoration/annotation that could optionally be provided with table data to specify how it would like to be rendered. Sadly I don't have any Nu fu to offer here as I'm new.
